api/rest/controller_test.go

```diff
@@ -0,0 +1,434 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package rest_test
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"github.com/labstack/echo/v4"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/api/rest"
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+func TestController_GetRegister(t *testing.T) {
+	const (
+		keyHex   = "746573744b6579"
+		value    = "testValue"
+		valueHex = "7465737456616c7565"
+	)
+
+	tests := map[string]struct {
+		key         string
+		heightParam string
+		lastHeight  uint64
+
+		stateGet func([]byte) ([]byte, error)
+
+		wantStatus   int
+		wantResponse *rest.RegisterResponse
+		wantErr      assert.ErrorAssertionFunc
+	}{
+		"nominal case with heightParam": {
+			key:         keyHex,
+			heightParam: "425",
+			lastHeight:  312,
+
+			stateGet: func(bytes []byte) ([]byte, error) {
+				return []byte(value), nil
+			},
+
+			wantStatus: http.StatusOK,
+			wantResponse: &rest.RegisterResponse{
+				Height: 425,
+				Key:    keyHex,
+				Value:  valueHex,
+			},
+			wantErr: assert.NoError,
+		},
+		"nominal case using last height": {
+			key:        keyHex,
+			lastHeight: 312,
+
+			stateGet: func(bytes []byte) ([]byte, error) {
+				return []byte(value), nil
+			},
+
+			wantStatus: http.StatusOK,
+			wantResponse: &rest.RegisterResponse{
+				Height: 312,
+				Key:    keyHex,
+				Value:  valueHex,
+			},
+			wantErr: assert.NoError,
+		},
+		"invalid key in ctx parameters": {
+			key: "not-hexadecimal",
+
+			wantStatus: http.StatusBadRequest,
+			wantErr:    assert.Error,
+		},
+		"invalid heightParam (negative value)": {
+			key:         keyHex,
+			heightParam: "not a number",
+
+			wantStatus: http.StatusBadRequest,
+			wantErr:    assert.Error,
+		},
+		"key not found": {
+			key:        keyHex,
+			lastHeight: 312,
+
+			stateGet: func(bytes []byte) ([]byte, error) {
+				return nil, dps.ErrNotFound
+			},
+
+			wantStatus: http.StatusNotFound,
+			wantErr:    assert.Error,
+		},
+		"internal state error": {
+			key:        keyHex,
+			lastHeight: 312,
+
+			stateGet: func(bytes []byte) ([]byte, error) {
+				return nil, errors.New("dummy error")
+			},
+
+			wantStatus: http.StatusInternalServerError,
+			wantErr:    assert.Error,
+		},
+	}
+
+	for desc, test := range tests {
+		test := test
+		t.Run(desc, func(t *testing.T) {
+			t.Parallel()
+
+			// Forge echo context and insert parameters.
+			e := echo.New()
+
+
+			u, err := url.Parse(fmt.Sprintf("https://1.2.3.4/values/:keys"))
+			require.NoError(t, err)
```

Comment:
Would be neat indeed. I usually use `require` for things that are not specifically related to the package that is being tested, as it provides clearer outputs than panicking. What else would you suggest (since there is no `url.MustParse`?